docker/android/android/soong.py

```diff
@@ -0,0 +1,654 @@
+              option: false,
+            },
+          },
+        }
+
+    the specification can be found below:
+    https://source.android.com/docs/core/tests/development/blueprints
+    https://android.googlesource.com/platform/build/soong/+/refs/heads/master/README.md
+
+    they also support single-line C++-style and multiline C-style comments.
+    the valid types are:
+        - bool (`true`, `false`)
+        - int
+        - string
+        - list (of strings)
+        - map
+
+    both lists and maps support optional trailing commas. any value type
+    can be present in a map, while only strings are allowed in lists.
+    integers, strings, arrays and maps also also support the `+` operator,
+    where `+` sums up integers. for strings and arrays, it appends the new
+    data. for maps, it produces the union of both keys, and for keys present
+    in both, it appends the value on the right-operand to the value in the
+    left one.
+
+    variable assignment produces immutable types, except for the `+=` operator.
+    `+=` does the described operation above in-place.
+
+    this parser doesn't need to be exactly correct: it does not need to reject
+    subtley invalid input. for example `name = {  }` may or may not be correct,
+    but it's fine to accept it as long as we output it identically. this is
+    supposed to handle all correct input and outputs it as correct output:
+    it doesn't need to validate type correctness.
+
+    this uses LALR parsing since it makes the grammar very easy to define and
+    the parsing simple. since the build step and repository synchronization
+    is much slower, the performance here is practically irrelevant.
+'''
+
+import json
+import sys
+
+import sly
+
+from . import util
+
+# dictionaries got insertion order in 3.6, guaranteed in 3.7
+assert sys.version_info >= (3, 6)
+
+# base character defs
+_H = r'[0-9a-f]'
+_NL = r'\n|\r\n|\r|\f'
+_UNICODE = fr'\\{_H}{1,6}(\r\n|[ \t\r\n\f])?'
+_ESCAPE = r'{_UNICODE}|\\[^\r\n\f0-9a-f]'
+_SINGLELINE_COMMENT = r'\/\/.*'
+# can't use reflags without setting them for all, so do manual dotall
+_MULTILINE_COMMENT = r'\/\*[\u0000-\U0010FFFF]*?\*\/'
+_COMMENT = fr'(?:{_SINGLELINE_COMMENT})|(?:{_MULTILINE_COMMENT})'
+
+
+def loads(contents, *_, **__):
+    return Ast.loads(contents)
+
+
+def load(fp, *_, **__):
+    return Ast.load(fp)
+
+
+def dumps(soong, pretty=True, indent=4, *_, **__):
+    return soong.dumps(pretty, indent)
+
+
+def dump(soong, fp, pretty=True, indent=4, *_, **__):
+    return soong.dump(fp, pretty, indent)
+
+
+class Lexer(sly.Lexer):
+    tokens = {
+        BOOL,
+        INTEGER,
+        IDENT,
+        STRING,
+        LBRACKET,
+        RBRACKET,
+        LBRACE,
+        RBRACE,
+        COLON,
+        COMMA,
+        EQUALS,
+        PLUS,
+    }
+    ignore = ' \t'
+    ignore_comment = _COMMENT
+
+    # Tokens
+    # this uses a string regex based on the CSS2.1 grammar
+    STRING = fr'"([^\n\r\f\\"]|\\{_NL}|{_ESCAPE})*"'
```

Comment:
Our complex grammars are mostly borrowed for the CSS2 specifications (before they swapped to railroad diagrams), which are great because it fully conforms to the Soong grammar.